[motion-path] Use border radius for coord-box

nmoucht opened the pull request:
#### 2f9a2a0138750b5ab4bd34fc487766a4adfb4c66
<pre>
[motion-path] Use border radius for coord-box
<a href="https://bugs.webkit.org/show_bug.cgi?id=261074">https://bugs.webkit.org/show_bug.cgi?id=261074</a>
rdar://114883507

Reviewed by Tim Nguyen.

Change containingBlockBoundingRect to FloatRoundedRect to take into account the border-radius
property.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-coord-box-001.html:
* Source/WebCore/rendering/MotionPath.cpp:
(WebCore::MotionPath::motionPathDataForRenderer):
(WebCore::MotionPath::lengthForRayPath):
(WebCore::MotionPath::computePathForRay):
(WebCore::MotionPath::computePathForBox):
* Source/WebCore/rendering/MotionPath.h:
* Source/WebCore/rendering/PathOperation.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/267613@main">https://commits.webkit.org/267613@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/afadbe94a3b6339f7d94d05f4d7959d0fb263364

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17187 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17513 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18009 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18975 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16086 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17380 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20787 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17655 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18281 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17391 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17734 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14926 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19792 "Built successfully") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14974 "1 flakes") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15617 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/22304 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15973 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15784 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20122 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16375 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/13892 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15528 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/15457 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4099 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19896 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16206 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/MotionPath.cpp:
```diff
@@ -53,20 +53,11 @@ std::optional<MotionPathData> MotionPath::motionPathDataForRenderer(const Render
-            data.containingBlockBoundingRect = snapRectToDevicePixelsIfNeeded(container->referenceBoxRect(shapePathOperation.referenceBox()), downcast<RenderLayerModelObject>(renderer));
-            data.offsetFromContainingBlock = offsetFromContainer(renderer, *container, data.containingBlockBoundingRect);
+            auto referenceRect = snapRectToDevicePixelsIfNeeded(container->referenceBoxRect(pathOperation->referenceBox()), downcast<RenderLayerModelObject>(renderer));
+            data.containingBlockBoundingRect = FloatRoundedRect(container->style().getRoundedBorderFor(LayoutRect(referenceRect)));
```
The spec says:
> If `<offset-path>` is omitted, it defaults to inset(0 round X), where X is the value of [border-radius](https://drafts.csswg.org/css-backgrounds-3/#propdef-border-radius) on the element that establishes the [containing block](https://drafts.csswg.org/css-display-4/#containing-block) for this element. If [<coord-box>](https://drafts.csswg.org/css-box-4/#typedef-coord-box) is omitted, it defaults to border-box.

Can we add WPTs that visually this does not affect ShapePathOperation? I know it doesn't right now, but it would be good still to prevent potential regressions.

nt1m replied:
(I'm also thinking maybe we should be intentional about only passing the border-radius in the `BoxPathOperation` case, since we're intentionally not passing it for `RayPathOperation`)

nmoucht replied:
Added. Also refactored the code to only pass for BoxPathOperation.

nt1m commented on Source/WebCore/rendering/MotionPath.cpp:
```diff
@@ -81,8 +72,7 @@ std::optional<MotionPathData> MotionPath::motionPathDataForRenderer(const Render
             data.offsetFromContainingBlock = offsetFromContainer(renderer, *container, pathReferenceBoxRect);
-            data.containingBlockBoundingRect = pathReferenceBoxRect;
-
+            data.containingBlockBoundingRect = FloatRoundedRect(pathReferenceBoxRect);
```
Can we avoid duplicating the code that computes the pathReferenceBoxRect ?

nmoucht replied:
Refactored.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-coord-box-001.html:
Please don't forget to export his change!

nmoucht committed:
[motion-path] Use border radius for coord-box
https://bugs.webkit.org/show_bug.cgi?id=261074
rdar://114883507

Reviewed by Tim Nguyen.

Change containingBlockBoundingRect to FloatRoundedRect to take into account the border-radius
property.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-coord-box-001.html:
* Source/WebCore/rendering/MotionPath.cpp:
(WebCore::MotionPath::motionPathDataForRenderer):
(WebCore::MotionPath::lengthForRayPath):
(WebCore::MotionPath::computePathForRay):
(WebCore::MotionPath::computePathForBox):
* Source/WebCore/rendering/MotionPath.h:
* Source/WebCore/rendering/PathOperation.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/267613@main
